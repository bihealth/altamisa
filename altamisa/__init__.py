# -*- coding: utf-8 -*-
"""Top-level package for AltamISA."""

__author__ = 'Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>'
# TODO: add version with versioneer
