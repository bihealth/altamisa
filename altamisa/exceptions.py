# -*- coding: utf-8 -*-
"""
Exceptions and Warnings used in the AltamISA library.
"""

__author__ = "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>"


class IsaException(Exception):
    """Base class for exceptions raised by Altamisa."""


class ParseIsatabException(IsaException):
    """Exception raised on problems parsing ISA-TAB."""


class WriteIsatabException(IsaException):
    """Exception raised on problems writing ISA-TAB."""


class IsaWarning(Warning):
    """Base class for warnings raised by Altamisa."""


class ParseIsatabWarning(IsaWarning):
    """Warning raised on problems parsing ISA-TAB."""


class WriteIsatabWarning(IsaWarning):
    """Warning raised on problems writing ISA-TAB."""


class IsaValidationWarning(IsaWarning):
    """Warning raised on problems validating ISA models or objects."""


class AdvisoryIsaValidationWarning(IsaValidationWarning):
    """Warning raised on uncritical problems when validating ISA models or objects."""


class ModerateIsaValidationWarning(IsaValidationWarning):
    """Warning raised on moderate problems when validating ISA models or objects."""


class CriticalIsaValidationWarning(IsaValidationWarning):
    """Warning raised on critical problems when validating ISA models or objects."""
