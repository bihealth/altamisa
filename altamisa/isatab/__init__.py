# -*- coding: utf-8 -*-
"""Module for parsing ISA-Tab files."""

# Make all models and the ``*Reader`` classes visible within this module.

from .headers import *  # noqa: F403, F401
from .models import *  # noqa: F403, F401
from .parse_assay_study import (  # noqa: F401
    AssayReader,
    AssayRowReader,
    StudyReader,
    StudyRowReader,
)
from .parse_investigation import InvestigationReader  # noqa: F401
from .validate_assay_study import AssayValidator, StudyValidator  # noqa: F401
from .validate_investigation import InvestigationValidator  # noqa: F401
from .write_assay_study import AssayWriter, StudyWriter, RefTableBuilder  # noqa: F401
from .write_investigation import InvestigationWriter  # noqa: F401
