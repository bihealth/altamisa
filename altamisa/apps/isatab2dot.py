# -*- coding: utf-8 -*-
"""Conversion of ISA-Tab to dot.
"""

import json
import sys
import os
import argparse

from altamisa.isatab import InvestigationReader, StudyReader, AssayReader


def print_dot(
    obj,
    outf,
    indent="    ",
    mat_shape="box",
    mat_color="black",
    proc_shape="ellipse",
    proc_color="blue",
):

    print(indent + "/* materials */", file=outf)
    for name, mat in obj.materials.items():
        label = json.dumps("{}:\n{}\n({})".format(mat.type, mat.name if mat.name else "-", name))
        print(
            "{}{} [label={},shape={},color={},fontcolor={}]".format(
                indent, json.dumps(name), label, mat_shape, mat_color, mat_color
            ),
            file=outf,
        )
    print(indent + "/* processes */", file=outf)
    for name, proc in obj.processes.items():
        label = json.dumps(
            "{}:\n{}\n{}\n({})".format(
                "Process",
                proc.protocol_ref if proc.protocol_ref else "-",
                proc.name if proc.name else "-",
                name,
            )
        )
        print(
            "{}{} [label={},shape={},color={},fontcolor={}]".format(
                indent, json.dumps(name), label, proc_shape, proc_color, proc_color
            ),
            file=outf,
        )
    print(indent + "/* arcs */", file=outf)
    for arc in obj.arcs:
        print("{}{} -> {};".format(indent, json.dumps(arc.tail), json.dumps(arc.head)), file=outf)


def run(args):
    with open(args.investigation_file, "rt") as inputf:
        investigation = InvestigationReader.from_stream(inputf).read()

    path = os.path.dirname(args.investigation_file)

    print("digraph investigation {", file=args.output_file)
    print('  rankdir = "LR";', file=args.output_file)

    for s, study_info in enumerate(investigation.studies):
        with open(os.path.join(path, study_info.info.path), "rt") as inputf:
            study = StudyReader.from_stream("S{}".format(s + 1), inputf).read()
        print("  /* study {} */".format(study_info.info.path), file=args.output_file)
        print("  subgraph clusterStudy{} {{".format(s), file=args.output_file)
        print('    label = "Study: {}"'.format(study_info.info.path), file=args.output_file)
        print_dot(study, args.output_file)
        print("  }", file=args.output_file)

        for a, assay_info in enumerate(study_info.assays):
            with open(os.path.join(path, assay_info.path), "rt") as inputf:
                assay = AssayReader.from_stream(
                    "S{}".format(s + 1), "A{}".format(a + 1), inputf
                ).read()
            print("  /* assay {} */".format(assay_info.path), file=args.output_file)
            print("  subgraph clusterAssayS{}A{} {{".format(s, a), file=args.output_file)
            print('    label = "Assay: {}"'.format(assay_info.path), file=args.output_file)
            print_dot(assay, args.output_file)
            print("  }", file=args.output_file)

    print("}", file=args.output_file)


def main(argv=None):
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-i", "--investigation-file", required=True, help="Path to investigation file"
    )
    parser.add_argument(
        "-o",
        "--output-file",
        default="-",
        type=argparse.FileType("wt"),
        help='Path to output file, stdout ("-") by default',
    )

    args = parser.parse_args(argv)
    return run(args)


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover
