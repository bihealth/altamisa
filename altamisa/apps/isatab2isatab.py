# -*- coding: utf-8 -*-
"""Read from ISA-Tab and directly write to ISA-Tab.
"""

import argparse
import os
import sys
import warnings

from altamisa.isatab import (
    AssayReader,
    AssayValidator,
    AssayWriter,
    InvestigationReader,
    InvestigationValidator,
    InvestigationWriter,
    StudyReader,
    StudyValidator,
    StudyWriter,
)
from altamisa.exceptions import IsaException


def run(args):
    # Collect warnings
    with warnings.catch_warnings(record=True) as records:
        run_warnings_caught(args)

    # Print warnings
    if not args.no_warnings:
        for record in records:
            warnings.showwarning(
                record.message, record.category, record.filename, record.lineno, record.line
            )


def run_warnings_caught(args):
    # Check if input and output directory are different
    path_in = os.path.realpath(os.path.dirname(args.input_investigation_file))
    path_out = os.path.realpath(os.path.dirname(args.output_investigation_file))
    if path_in == path_out:
        tpl = "Can't output ISA-tab files to same directory as as input: {} == {}"
        msg = tpl.format(path_in, path_out)
        raise IsaException(msg)

    if args.input_investigation_file == "-":  # pragma: no cover
        args.input_investigation_file = sys.stdin
    else:
        args.input_investigation_file = open(args.input_investigation_file, "rt")
    if args.output_investigation_file == "-":  # pragma: no cover
        args.output_investigation_file = sys.stdout
    else:
        args.output_investigation_file = open(args.output_investigation_file, "wt")

    investigation, studies, assays = run_reading(args, path_in)
    run_writing(args, path_out, investigation, studies, assays)


def run_reading(args, path_in):
    # Read investigation
    investigation = InvestigationReader.from_stream(args.input_investigation_file).read()

    # Validate investigation
    InvestigationValidator(investigation).validate()

    # Read studies and assays
    studies = {}
    assays = {}
    for s, study_info in enumerate(investigation.studies):
        if study_info.info.path:
            with open(os.path.join(path_in, study_info.info.path), "rt") as inputf:
                studies[s] = StudyReader.from_stream("S{}".format(s + 1), inputf).read()
        if study_info.assays:
            assays[s] = {}
        for a, assay_info in enumerate(study_info.assays):
            if assay_info.path:
                with open(os.path.join(path_in, assay_info.path), "rt") as inputf:
                    assays[s][a] = AssayReader.from_stream(
                        "S{}".format(s + 1), "A{}".format(a + 1), inputf
                    ).read()

    # Validate studies and assays
    for s, study_info in enumerate(investigation.studies):
        if study_info.info.path:
            StudyValidator(investigation, study_info, studies[s]).validate()
        for a, assay_info in enumerate(study_info.assays):
            if assay_info.path:
                AssayValidator(investigation, study_info, assay_info, assays[s][a]).validate()

    return investigation, studies, assays


def run_writing(args, path_out, investigation, studies, assays):
    # Write investigation
    if args.output_investigation_file.name == "<stdout>":
        InvestigationWriter.from_stream(
            investigation, args.output_investigation_file, quote=args.quotes
        ).write()
    else:
        with open(args.output_investigation_file.name, "wt", newline="") as outputf:
            InvestigationWriter.from_stream(investigation, outputf, quote=args.quotes).write()

    # Write studies and assays
    for s, study_info in enumerate(investigation.studies):
        if args.output_investigation_file.name == "<stdout>":
            if study_info.info.path:
                StudyWriter.from_stream(
                    studies[s], args.output_investigation_file, quote=args.quotes
                ).write()
            for a, assay_info in enumerate(study_info.assays):
                if assay_info.path:
                    AssayWriter.from_stream(
                        assays[s][a], args.output_investigation_file, quote=args.quotes
                    ).write()
        else:
            if study_info.info.path:
                with open(
                    os.path.join(path_out, study_info.info.path), "wt", newline=""
                ) as outputf:
                    StudyWriter.from_stream(studies[s], outputf, quote=args.quotes).write()
            for a, assay_info in enumerate(study_info.assays):
                if assay_info.path:
                    with open(os.path.join(path_out, assay_info.path), "wt", newline="") as outputf:
                        AssayWriter.from_stream(assays[s][a], outputf, quote=args.quotes).write()


def main(argv=None):
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-i",
        "--input-investigation-file",
        required=True,
        type=str,
        help="Path to input investigation file",
    )
    parser.add_argument(
        "-o",
        "--output-investigation-file",
        default="-",
        type=str,
        help=(
            'Path to output investigation file, stdout ("-") by default. '
            "Needs to be in a different directory!"
        ),
    )
    parser.add_argument(
        "-q",
        "--quotes",
        default=None,
        type=str,
        help='Character for quoting, e.g. "\\"" (None by default)',
    )
    parser.add_argument(
        "--no-warnings",
        dest="no_warnings",
        action="store_true",
        help="Suppress ISA-tab related warnings (False by default)",
    )
    parser.set_defaults(no_warnings=False)

    args = parser.parse_args(argv)
    return run(args)


if __name__ == "__main__":
    sys.exit(main())  # pragma: no cover
