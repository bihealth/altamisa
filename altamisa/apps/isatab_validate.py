# -*- coding: utf-8 -*-
"""Read from ISA-Tab and print validation warnings, if any.
"""

import argparse
import os
import sys
import warnings

from altamisa.isatab import (
    AssayReader,
    AssayValidator,
    InvestigationReader,
    InvestigationValidator,
    StudyReader,
    StudyValidator,
)


def run(args):
    # Show all warnings of same type and content
    if args.show_duplicate_warnings:
        warnings.simplefilter("always")

    # Collect warnings
    with warnings.catch_warnings(record=True) as records:
        run_warnings_caught(args)

    # Print warnings
    for record in records:
        warnings.showwarning(
            record.message, record.category, record.filename, record.lineno, record.line
        )


def run_warnings_caught(args):
    # Read investigation
    investigation = InvestigationReader.from_stream(args.input_investigation_file).read()
    args.input_investigation_file.close()

    # Validate investigation
    InvestigationValidator(investigation).validate()

    # Read studies and assays
    path_in = os.path.normpath(os.path.dirname(args.input_investigation_file.name))
    studies = {}
    assays = {}
    for s, study_info in enumerate(investigation.studies):
        if study_info.info.path:
            with open(os.path.join(path_in, study_info.info.path), "rt") as inputf:
                studies[s] = StudyReader.from_stream("S{}".format(s + 1), inputf).read()
        if study_info.assays:
            assays[s] = {}
        for a, assay_info in enumerate(study_info.assays):
            if assay_info.path:
                with open(os.path.join(path_in, assay_info.path), "rt") as inputf:
                    assays[s][a] = AssayReader.from_stream(
                        "S{}".format(s + 1), "A{}".format(a + 1), inputf
                    ).read()

    # Validate studies and assays
    for s, study_info in enumerate(investigation.studies):
        if study_info.info.path:
            StudyValidator(investigation, study_info, studies[s]).validate()
        for a, assay_info in enumerate(study_info.assays):
            if assay_info.path:
                AssayValidator(
                    investigation, study_info, assay_info, assays[s][a], studies[s]
                ).validate()


def main(argv=None):
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-i",
        "--input-investigation-file",
        required=True,
        type=argparse.FileType("rt"),
        help="Path to input investigation file",
    )
    parser.add_argument(
        "--show-duplicate-warnings",
        dest="show_duplicate_warnings",
        action="store_true",
        help=(
            "Show duplicated warnings, i.e. with same message and same category (False by default)"
        ),
    )
    parser.set_defaults(no_warnings=False)

    args = parser.parse_args(argv)
    return run(args)


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main())
