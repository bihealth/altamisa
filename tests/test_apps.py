# -*- coding: utf-8 -*-
"""Smoke tests for the apps"""

import os.path

import pytest
from syrupy.assertion import SnapshotAssertion
from typer.testing import CliRunner

from altamisa.apps import isatab2dot, isatab2isatab, isatab_validate
from altamisa.exceptions import IsaWarning

runner = CliRunner()


def test_isatab_validate(snapshot: SnapshotAssertion):
    i_file = os.path.join(os.path.dirname(__file__), "data", "i_warnings", "i_warnings.txt")
    argv = ["--input-investigation-file", i_file, "--show-duplicate-warnings"]

    with pytest.warns(IsaWarning) as record:
        result = runner.invoke(isatab_validate.app, argv)
        assert result.exit_code == 0

    assert snapshot == [str(r.message) for r in record]


def test_isatab2isatab(tmpdir, snapshot: SnapshotAssertion):
    i_file = os.path.join(os.path.dirname(__file__), "data", "i_minimal", "i_minimal.txt")
    argv = [
        "--input-investigation-file",
        i_file,
        "--output-investigation-file",
        str(tmpdir.mkdir("i_minimal").join("i_minimal.txt")),
        "--quotes",
        '"',
    ]

    with pytest.warns(IsaWarning) as record:
        result = runner.invoke(isatab2isatab.app, argv)
        assert result.exit_code == 0

    assert snapshot == [str(r.message) for r in record]


def test_isatab2isatab_input_is_output(tmpdir, snapshot: SnapshotAssertion):
    i_file = os.path.join(os.path.dirname(__file__), "data", "i_minimal", "i_minimal.txt")
    argv = [
        "--input-investigation-file",
        i_file,
        "--output-investigation-file",
        i_file,
        "--quotes",
        '"',
    ]

    result = runner.invoke(isatab2isatab.app, argv)
    assert result.exit_code == 1
    assert snapshot == str(result)


def test_isatab2dot(tmpdir):
    i_file = os.path.join(os.path.dirname(__file__), "data", "i_minimal", "i_minimal.txt")
    argv = [
        "--investigation-file",
        i_file,
        "--output-file",
        str(tmpdir.mkdir("dot").join("out.dot")),
    ]

    result = runner.invoke(isatab2dot.app, argv)
    assert result.exit_code == 0
