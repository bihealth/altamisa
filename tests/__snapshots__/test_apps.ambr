# serializer version: 1
# name: test_isatab2isatab
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	i_minimal.txt
      	Submission Date:	
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
    'No reference headers available for section INVESTIGATION PUBLICATIONS. Applying default order.',
    'No reference headers available for section INVESTIGATION CONTACTS. Applying default order.',
    'No reference headers available for section STUDY DESIGN DESCRIPTORS. Applying default order.',
    'No reference headers available for section STUDY PUBLICATIONS. Applying default order.',
    'No reference headers available for section STUDY FACTORS. Applying default order.',
    'No reference headers available for section STUDY CONTACTS. Applying default order.',
  ])
# ---
# name: test_isatab2isatab_input_is_output
  '<Result IsaException("Can\'t output ISA-tab files to same directory as as input: /home/sellts_c/Code/altamisa/tests/data/i_minimal == /home/sellts_c/Code/altamisa/tests/data/i_minimal")>'
# ---
# name: test_isatab_validate
  list([
    'Incomplete ontology source; found: , Incomplete 1, 1, Incomplete 1, ()',
    'Incomplete ontology source; found: Incomplete 2, , 2, Incomplete 2, ()',
    'Ontology source name including whitespace(s); found: Incomplete 2, , 2, Incomplete 2, ()',
    'Incomplete ontology source; found: Incomplete 3, Incomplete 3, , Incomplete 3, ()',
    'Ontology source name including whitespace(s); found: Incomplete 3, Incomplete 3, , Incomplete 3, ()',
    'Incomplete ontology source; found: Incomplete 4, Incomplete 4, 4, , ()',
    'Ontology source name including whitespace(s); found: Incomplete 4, Incomplete 4, 4, , ()',
    '''
      Investigation with only one study contains metadata:
      	ID:	i_warnings
      	Title:	Investigation with Warnings
      	Path:	i_warnings.txt
      	Submission Date:	
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    'Invalid mail address: invalid_mail',
    'Invalid phone/fax number: CALL-ME',
    'Invalid phone/fax number: FAX-ME',
    'Invalid pubmed_id string: not-pubmed',
    'Invalid doi string: not-a-doi',
    '''
      Assay without platform:
      Path:	a_warnings.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
    'Assay path used more than once: a_warnings.txt',
    "Found samples in assay 'a_warnings.txt' but not in parent study 's_warnings.txt':\\n0815-N2",
    "Found samples in assay 'a_warnings.txt' but not in parent study 's_warnings.txt':\\n0815-N2",
  ])
# ---
